mlir/include/Catalyst/Driver/CompilerDriver.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <filesystem>
+#include <vector>
+
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
+
+/// Data about the JIT function that is optionally inferred and returned to the caller.
+///
+/// This is important for calling a function when invoking the compiler on an MLIR or LLVM textual
+/// representation intead of from Python.
+struct FunctionAttributes {
+    /// The name of the primary JIT entry point function.
+    std::string functionName;
+    /// The return type of the JIT entry point function.
+    std::string returnType;
+};
+
+/// Verbosity level
+// TODO: Adjust the number of levels according to our needs. MLIR seems to print few really
+// low-level messages, we might want to hide these.
+enum class Verbosity { Silent = 0, Urgent = 1, Debug = 2, All = 3 };
+
+/// Helper verbose reporting macro.
+#define CO_MSG(opt, level, op)                                                                     \
+    do {                                                                                           \
+        if ((opt).verbosity >= (level)) {                                                          \
+            (opt).diagnosticStream << op;                                                          \
+        }                                                                                          \
+    } while (0)
```

Comment:
[See here](https://hownot2code.wordpress.com/2016/12/05/do-while-0-in-macros/)

Reply:
To be fair, I also would prefer to just write this logging as a simple function. The only time I prefer macros over functions is when I have a macro that outputs line information via `__LINE__` and similar other macros. I am not sure if there is an advantage to keeping it as a macro if this hypothetical function were to be inlined.

Reply:
I do see the advantage now, since `op` only needs to define the `<<` operator and does not need to be typed. If we were to have a function, it would need to be a template due to `op`. I think we should keep it as is.

Reply:
I know the syntax error related to C macros :)  but it shouldn't be the problem with this if condition. we don't have the else branch in the macro that cause an issue. Anyway, this is good! don't worry about it.